File: backend/app/schemas/asset.py
import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

# 前端 AssetType 定义（与 frontend/src/lib/types.ts 对齐）
AssetTypeValue = Literal["character", "scene", "object", "style", "audio", "other"]


# --- Folder ---

class AssetFolderCreate(BaseModel):
    name: str
    space_key: str = "personal"
    parent_id: Optional[int] = None


class AssetFolderUpdate(BaseModel):
    name: Optional[str] = None


class AssetFolderOut(BaseModel):
    id: int
    user_id: int
    name: str
    space_key: str
    parent_id: Optional[int]
    created_at: datetime.datetime

    model_config = {"from_attributes": True}


# --- Asset ---

class AssetCreate(BaseModel):
    name: str = Field(default="Untitled", max_length=200)
    type: AssetTypeValue = "other"
    width: int = 0
    height: int = 0
    description: str = ""
    tags: list[str] = []
    extra_data: dict = {}
    folder_id: Optional[int] = None
    space_key: str = "personal"


class AssetUpdate(BaseModel):
    name: Optional[str] = Field(default=None, max_length=200)
    type: Optional[AssetTypeValue] = None
    width: Optional[int] = None
    height: Optional[int] = None
    description: Optional[str] = None
    tags: Optional[list[str]] = None
    folder_id: Optional[int] = None


class AssetOut(BaseModel):
    id: int
    user_id: int
    folder_id: Optional[int]
    space_key: str
    name: str
    type: str
    width: int
    height: int
    description: str
    tags: list
    extra_data: dict
    created_at: datetime.datetime
    updated_at: datetime.datetime

    model_config = {"from_attributes": True}

File: server/schemas/asset.ts
/**
 * 资产相关请求校验模式。
 * 定义文件夹与资产的创建、更新等入参的 zod 校验规则。
 */
import { z } from "zod";

// Folder
export const folderCreateSchema = z.object({
  name: z.string().min(1).max(50),
  scope: z.literal("personal").optional(),
  parentId: z.number().int().positive().nullable().optional(),
});

export const folderUpdateSchema = z.object({
  name: z.string().min(1).max(50),
});

export const folderOutSchema = z.object({
  id: z.number(),
  userId: z.number(),
  name: z.string(),
  scope: z.string(),
  kind: z.string(),
  parentId: z.number().nullable(),
  createdAt: z.string(),
  count: z.number(),
});

export type FolderCreate = z.infer<typeof folderCreateSchema>;
export type FolderUpdate = z.infer<typeof folderUpdateSchema>;
export type FolderOut = z.infer<typeof folderOutSchema>;

// Asset Item
export const assetCreateSchema = z.object({
  name: z.string().min(1).max(200).optional(),
  type: z.string().max(20).optional(),
  mediaType: z.string().max(10).optional(),
  sourceUrl: z.string().min(1).optional(),
  sourceType: z.string().max(20).optional(),
  width: z.number().int().min(0).optional(),
  height: z.number().int().min(0).optional(),
  description: z.string().optional(),
  tags: z.array(z.string()).optional(),
  extraData: z.record(z.unknown()).optional(),
  folderId: z.number().int().positive().nullable().optional(),
  scope: z.literal("personal").optional(),
});

export const assetBatchCreateSchema = z.array(assetCreateSchema).min(1).max(200);

export const assetUpdateSchema = z.object({
  name: z.string().min(1).max(200).optional(),
  type: z.string().min(1).max(20).optional(),
  mediaType: z.string().max(10).optional(),
  width: z.number().int().min(0).optional(),
  height: z.number().int().min(0).optional(),
  description: z.string().optional(),
  folderId: z.number().int().positive().nullable().optional(),
  tags: z.array(z.string()).optional(),
  extraData: z.record(z.unknown()).optional(),
});

export const assetBatchUpdateSchema = z.object({
  ids: z.array(z.number().int().positive()).min(1).max(200),
  updates: z.object({
    folderId: z.number().int().positive().nullable().optional(),
    type: z.string().min(1).max(20).optional(),
  }),
});

export const assetOutSchema = z.object({
  id: z.number(),
  userId: z.number(),
  folderId: z.number(),
  scope: z.string(),
  sourceUrl: z.string().nullable(),
  sourceType: z.string(),
  name: z.string(),
  type: z.string(),
  mediaType: z.string(),
  width: z.number(),
  height: z.number(),
  description: z.string(),
  tags: z.array(z.string()),
  extraData: z.record(z.unknown()),
  createdAt: z.string(),
  updatedAt: z.string(),
});

export type AssetCreate = z.infer<typeof assetCreateSchema>;
export type AssetUpdate = z.infer<typeof assetUpdateSchema>;
export type AssetBatchUpdate = z.infer<typeof assetBatchUpdateSchema>;
export type AssetOut = z.infer<typeof assetOutSchema>;
